### PR TITLE
Field effect

### PR DESCRIPTION
**Pull request by PikalaxALT**

### CODE REVIEW

File: src/fldeff_rocksmash.c [hunk 1]
```diff
@@ -48,7 +48,7 @@ static void Task_FieldEffectShowMon_Init(u8 taskId)
     u8 mapObjId;
 
     ScriptContext2_Enable();
-    gPlayerAvatar.unk6 = TRUE;
+    gPlayerAvatar.preventStep = TRUE;
```
Comment (jiangzhengwenjz) [inline src/fldeff_rocksmash.c]: So this field should be `bool8` when declared

File: src/fieldmap.c [hunk 2]
```diff
@@ -946,7 +946,7 @@ void sub_8059948(u8 a0, u8 a1)
             TintPalette_SepiaTone(gPlttBufferUnfaded + a0 * 16, a1 * 16);
             break;
         case 3:
-            sub_8111F38(a0 * 16, a1 * 16);
+            sub_8111F38((a0 * 16), a1 * 16);
```
Comment (jiangzhengwenjz) [inline src/fieldmap.c]: What's the modification for?

File: src/vs_seeker.c [hunk 3]
```diff
@@ -748,7 +748,7 @@ void Task_VsSeeker_0(u8 taskId)
     else if (respval == 2)
     {
         ItemUse_SetQuestLogEvent(QL_EVENT_USED_ITEM, 0, gSpecialVar_ItemId, 0xffff);
-        FieldEffectStart(FLDEFF_UNK_41); // TODO: name this enum
+        FieldEffectStart(FLDEFF_USE_VS_SEEKER); // TODO: name this enum
```
Comment (jiangzhengwenjz) [inline src/vs_seeker.c]: remove comment

File: src/field_effect.c [hunk 4]
```diff
@@ -0,0 +1,3907 @@
...
+static const u16 gUnknown_83CAFB0[] = INCBIN_U16("graphics/field_effects/unk_83CAFB0.gbapal");
+static const u16 gUnknown_83CAFD0[] = INCBIN_U16("graphics/field_effects/unk_83CAFD0.4bpp");
+static const u16 gUnknown_83CB3D0[] = INCBIN_U16("graphics/field_effects/unk_83CB3D0.gbapal");
+static const u16 gUnknown_83CB3F0[] = INCBIN_U16("graphics/field_effects/unk_83CB3F0.4bpp");
+static const u16 gFieldMoveStreaksTiles[] = INCBIN_U16("graphics/field_effects/unk_83CB5F0.4bpp");
+static const u16 gFieldMoveStreaksPalette[] = INCBIN_U16("graphics/field_effects/unk_83CB7F0.gbapal");
+static const u16 gFieldMoveStreaksTilemap[] = INCBIN_U16("graphics/field_effects/unk_83CB810.bin");
+static const u16 gDarknessFieldMoveStreaksTiles[] = INCBIN_U16("graphics/field_effects/unk_83CBA90.4bpp");
+static const u16 gDarknessFieldMoveStreaksPalette[] = INCBIN_U16("graphics/field_effects/unk_83CBB10.gbapal");
+static const u16 gDarknessFieldMoveStreaksTilemap[] = INCBIN_U16("graphics/field_effects/unk_83CBB30.bin");
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: should start with `s`

File: src/field_effect.c [hunk 5]
```diff
@@ -0,0 +1,3907 @@
...
+    ANIMCMD_FRAME(0, 1),
+    ANIMCMD_END
+};
+
+static const union AnimCmd *const gNewGameOakAnimTable[] = {
+    gNewGameOakAnim
+};
+
+static const struct SpriteTemplate gNewGameOakObjectTemplate = {
+    .tileTag = 0xFFFF,
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `SPRITE_INVALID_TAG`

File: src/field_effect.c [hunk 6]
```diff
@@ -0,0 +1,3907 @@
...
+    ANIMCMD_FRAME(0, 8),
+    ANIMCMD_END
+};
+
+static const union AnimCmd *const gUnknown_83CBF84[] = {
+    gUnknown_83CBF5C
+};
+
+static const struct SpriteTemplate gUnknown_83CBF88 = {
+    .tileTag = 65535,
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `SPRITE_INVALID_TAG` (same comment below)

File: src/field_effect.c [hunk 7]
```diff
@@ -0,0 +1,3907 @@
...
+            return;
+    }
+    FreeSpriteTilesByTag(tileTag);
+}
+
+static void FieldEffectFreePaletteIfUnused(u8 paletteNum)
+{
+    u8 i;
+    u16 paletteTag = GetSpritePaletteTagByPaletteNum(paletteNum);
+    if (paletteTag == 0xFFFF)
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `SPRITE_INVALID_TAG`

File: src/field_effect.c [hunk 8]
```diff
@@ -0,0 +1,3907 @@
...
+    while (sLavaridgeGymB1FWarpEffectFuncs[gTasks[taskId].data[0]](&gTasks[taskId], &gObjectEvents[gPlayerAvatar.objectEventId], &gSprites[gPlayerAvatar.spriteId]));
+}
+
+static bool8 LavaridgeGymB1FWarpEffect_1(struct Task * task, struct ObjectEvent * objectEvent, struct Sprite * sprite)
+{
+    FreezeObjectEvents();
+    CameraObjectReset2();
+    SetCameraPanningCallback(NULL);
+    gPlayerAvatar.preventStep = TRUE;
+    objectEvent->fixedPriority = 1;
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE`

File: src/field_effect.c [hunk 9]
```diff
@@ -0,0 +1,3907 @@
...
+    return FALSE;
+}
+
+u8 FldEff_PopOutOfAsh(void)
+{
+    u8 spriteId;
+    sub_8063BC4((s16 *)&gFieldEffectArguments[0], (s16 *)&gFieldEffectArguments[1], 8, 8);
+    spriteId = CreateSpriteAtEnd(gFieldEffectObjectTemplatePointers[32], gFieldEffectArguments[0], gFieldEffectArguments[1], gFieldEffectArguments[2]);
+    gSprites[spriteId].oam.priority = gFieldEffectArguments[3];
+    gSprites[spriteId].coordOffsetEnabled = 1;
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE`

File: src/field_effect.c [hunk 10]
```diff
@@ -0,0 +1,3907 @@
...
+        task->data[3]++;
+        objectEvent->fixedPriority = 0;
+        sprite->subspriteMode = task->data[4];
+        objectEvent->triggerGroundEffectsOnMove = 1;
+    }
+    if (sprite->pos2.y >= 0)
+    {
+        PlaySE(SE_W070);
+        objectEvent->triggerGroundEffectsOnStop = 1;
+        objectEvent->landingJump = 1;
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE`

File: src/field_effect.c [hunk 11]
```diff
@@ -0,0 +1,3907 @@
...
+    return playerObj->facingDirection;
+}
+
+static bool32 sub_80857F0(struct ObjectEvent * playerObj, s16 *state_p, s16 *y_p)
+{
+    struct Sprite * sprite = &gSprites[playerObj->spriteId];
+    switch (*state_p)
+    {
+    case 0:
+        Camer
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `WIN_RANGE`

Comment (jiangzhengwenjz) [inline src/field_effect.c]: use constants

Comment (jiangzhengwenjz) [inline src/field_effect.c]: `(uintptr_t *)` (same comment below)

Comment (jiangzhengwenjz) [inline src/field_effect.c]: `WIN_RANGE`

Comment (jiangzhengwenjz) [inline src/field_effect.c]: use constants

Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE`

Comment (jiangzhengwenjz) [inline src/field_effect.c]: `FALSE`

Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE`

Comment (jiangzhengwenjz) [inline src/field_effect.c]: `FALSE`

File: src/fieldmap.c (hunk 2 above)
Comment (PikalaxALT) [reply src/fieldmap.c]: Incompletely reverted test change

File: src/field_effect.c [hunk 12]
```diff
@@ -0,0 +1,3907 @@
...
+{
+    FieldEffectFreeGraphicsResources(sprite);
+    FieldEffectActiveListRemove(fldeff);
+}
+
+static void FieldEffectFreeTilesIfUnused(u16 tileStart)
+{
+    u8 i;
+    u16 tileTag = GetSpriteTileTagByTileStart(tileStart);
+    if (tileTag == 0xFFFF)
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `SPRITE_INVALID_TAG`

File: src/field_effect.c [hunk 13]
```diff
@@ -0,0 +1,3907 @@
...
+    }
+    FreeSpriteTilesByTag(tileTag);
+}
+
+static void FieldEffectFreePaletteIfUnused(u8 paletteNum)
+{
+    u8 i;
+    u16 paletteTag = GetSpritePaletteTagByPaletteNum(paletteNum);
+    if (paletteTag == 0xFFFF)
+        return;
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `SPRITE_INVALID_TAG` 1 line above

File: src/field_effect.c [hunk 14]
```diff
@@ -0,0 +1,3907 @@
...
+
+// For the ash puff effect when warping off the B1F ash tiles
+u8 FldEff_LavaridgeGymWarp(void)
+{
+    u8 spriteId;
+    sub_8063BC4((s16 *)&gFieldEffectArguments[0], (s16 *)&gFieldEffectArguments[1], 8, 8);
+    spriteId = CreateSpriteAtEnd(gFieldEffectObjectTemplatePointers[33], gFieldEffectArguments[0], gFieldEffectArguments[1], gFieldEffectArguments[2]);
+    gSprites[spriteId].oam.priority = gFieldEffectArguments[3];
+    gSprites[spriteId].coordOffsetEnabled = 1;
+    return spriteId;
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE` 1 line above

File: src/field_effect.c [hunk 15]
```diff
@@ -0,0 +1,3907 @@
...
+}
+
+u8 FldEff_PopOutOfAsh(void)
+{
+    u8 spriteId;
+    sub_8063BC4((s16 *)&gFieldEffectArguments[0], (s16 *)&gFieldEffectArguments[1], 8, 8);
+    spriteId = CreateSpriteAtEnd(gFieldEffectObjectTemplatePointers[32], gFieldEffectArguments[0], gFieldEffectArguments[1], gFieldEffectArguments[2]);
+    gSprites[spriteId].oam.priority = gFieldEffectArguments[3];
+    gSprites[spriteId].coordOffsetEnabled = 1;
+    return spriteId;
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE` one line above

File: src/field_effect.c [hunk 16]
```diff
@@ -0,0 +1,3907 @@
...
+            (*stage_p)++;
+        *delay_p = 12 >> (*stage_p); // 12 >> 4 = 0
+        return sUnknown_83CC0E8[playerObj->facingDirection];
+    }
+    return playerObj->facingDirection;
+}
+
+static bool32 sub_80857F0(struct ObjectEvent * playerObj, s16 *state_p, s16 *y_p)
+{
+    struct Sprite * sprite = &gSprites[pla
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `TRUE` one line above

File: src/field_effect.c [hunk 17]
```diff
@@ -0,0 +1,3907 @@
...
+    FieldEffectFreeGraphicsResources(sprite);
+    FieldEffectActiveListRemove(fldeff);
+}
+
+static void FieldEffectFreeTilesIfUnused(u16 tileStart)
+{
+    u8 i;
+    u16 tileTag = GetSpriteTileTagByTileStart(tileStart);
+    if (tileTag == 0xFFFF)
+        return;
```
Comment (jiangzhengwenjz) [inline src/field_effect.c]: `SPRITE_INVALID_TAG` above

**Pull request merged**